daemon/containerd/image_exporter.go

```diff
@@ -71,3 +75,61 @@ func (i *ImageService) LoadImage(ctx context.Context, inTar io.ReadCloser, outSt
 	}
 	return nil
 }
+
+func (i *ImageService) appendImageForExport(ctx context.Context, opts []archive.ExportOpt, name string) ([]archive.ExportOpt, error) {
+	ref, err := reference.ParseDockerRef(name)
+	if err != nil {
+		return opts, err
+	}
+
+	is := i.client.ImageService()
+
+	img, err := is.Get(ctx, ref.String())
+	if err != nil {
+		return opts, err
+	}
+
+	store := i.client.ContentStore()
+
+	if containerdimages.IsIndexType(img.Target.MediaType) {
+		children, err := containerdimages.Children(ctx, store, img.Target)
+		if err != nil {
+			return opts, err
+		}
+
+		// Check which platform manifests we have blobs for.
+		missingPlatforms := []v1.Platform{}
+		presentPlatforms := []v1.Platform{}
+		for _, child := range children {
+			if containerdimages.IsManifestType(child.MediaType) {
+				_, err := store.ReaderAt(ctx, child)
+				if cerrdefs.IsNotFound(err) {
+					missingPlatforms = append(missingPlatforms, *child.Platform)
+					logrus.WithField("digest", child.Digest.String()).Debug("missing blob, not exporting")
+					continue
+				} else if err != nil {
+					return opts, err
+				}
+				presentPlatforms = append(presentPlatforms, *child.Platform)
+			}
+		}
+
+		// If we have all the manifests, just export the original index.
+		if len(missingPlatforms) == 0 {
+			return append(opts, archive.WithImage(is, img.Name)), nil
+		}
+
+		// Create a new manifest which contains only the manifests we have in store.
+		srcRef := ref.String()
+		targetRef := srcRef + "-tmp-export"
```

Comment:
@vvoland wondering if this needs a more unique name (in case two exports run simultaneously)

Reply:
Yes, definitely an unique name is needed here.

This is already done in https://github.com/rumpl/moby/pull/56/ which also has a bit better implementation of finding out which manifest are present in the store.
Maybe we could hold on a bit with this one and then squash changes so that we don't diverge too much?

Reply:
Ah, yes, I could combine both in this PR

Reply:
Didn't realize that one was also updating the `export` code 😅